[fuser] Support bfloat16

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54571 [fuser] Support bfloat16**

Supports bfloat16 via a similar method to half: upconvert inputs to
fp32, do math, then downconvert outputs to bf16.

Resource strings are mostly derived from cuda-11 headers.

Fixes #53918, for the legacy fuser at least.

Differential Revision: [D27328987](https://our.internmc.facebook.com/intern/diff/D27328987)

bertmaher committed:
[fuser] Support bfloat16

Supports bfloat16 via a similar method to half: upconvert inputs to
fp32, do math, then downconvert outputs to bf16.

Resource strings are mostly derived from cuda-11 headers.

[ghstack-poisoned]

bertmaher committed:
Update on "[fuser] Support bfloat16"


Supports bfloat16 via a similar method to half: upconvert inputs to
fp32, do math, then downconvert outputs to bf16.

Resource strings are mostly derived from cuda-11 headers.

Fixes #53918, for the legacy fuser at least.

[ghstack-poisoned]

ngimel reviewed (approved):
Awesome, thanks!

ngimel commented on torch/csrc/jit/codegen/fuser/codegen.cpp:
```diff
@@ -618,6 +640,9 @@ std::string generateKernel(
   }
+  if (has_bfloat_tensor) {
+    env.s("BFloat16Header", cuda::bfloat16_support_literal);
+  }
```
Test failures are due to this.

bertmaher committed:
Update on "[fuser] Support bfloat16"


Supports bfloat16 via a similar method to half: upconvert inputs to
fp32, do math, then downconvert outputs to bf16.

Resource strings are mostly derived from cuda-11 headers.

Fixes #53918, for the legacy fuser at least.

[ghstack-poisoned]
